governance/xc_admin/packages/xc_admin_frontend/components/programs/PythLazer.tsx

```diff
@@ -1,41 +1,503 @@
-import React from 'react'
+import React, { useState, useContext } from 'react'
+import { usePythContext } from '../../contexts/PythContext'
+import { ClusterContext } from '../../contexts/ClusterContext'
+import Modal from '../common/Modal'
+import {
+  validateUploadedConfig,
+  ProgramType,
+} from '@pythnetwork/xc-admin-common'
+import toast from 'react-hot-toast'
+import Loadbar from '../loaders/Loadbar'
+import Spinner from '../common/Spinner'
+import {
+  LazerState,
+  LazerFeed,
+  LazerPublisher,
+} from '@pythnetwork/xc-admin-common/src/programs/types'
+import { capitalizeFirstLetter } from '../../utils/capitalizeFirstLetter'
 
 interface PythLazerProps {
-  proposerServerUrl: string // kept for consistency with PythCore interface
+  proposerServerUrl: string
+}
+
+interface ShardChanges {
+  prev?: {
+    shardId: number
+    shardName: string
+    minRate: string
+  }
+  new: {
+    shardId: number
+    shardName: string
+    minRate: string
+  }
+}
+
+interface FeedChanges {
+  prev?: LazerFeed
+  new?: LazerFeed
+}
+
+interface PublisherChanges {
+  prev?: LazerPublisher
+  new?: LazerPublisher
+}
+
+interface ShardChangesRowsProps {
+  changes: ShardChanges
+}
+
+interface FeedChangesRowsProps {
+  changes: FeedChanges
+  feedId: string
+}
+
+interface PublisherChangesRowsProps {
+  changes: PublisherChanges
+  publisherId: string
+}
+
+interface ModalContentProps {
+  changes: Record<
+    string,
+    {
+      prev?: Partial<LazerState>
+      new?: Partial<LazerState>
+    }
+  >
+  onSendProposal: () => void
+  isSendProposalButtonLoading: boolean
+}
+
+const ShardChangesRows: React.FC<ShardChangesRowsProps> = ({ changes }) => {
+  const isNewShard = !changes.prev && changes.new
+
+  return (
+    <>
+      {Object.entries(changes.new).map(
+        ([key, newValue]) =>
+          (isNewShard ||
+            (changes.prev &&
+              changes.prev[key as keyof typeof changes.prev] !== newValue)) && (
```

Comment:
This is kind of awkward, if this is how you plan to use this then I'd rethink the type a bit.  The type system can't actually guaranteed that the keys are shared between `prev` and `new`, which is why you have the typecast.  So instead I think it would be better to have the type be something where you have `prev` and `new` underneath each key, i.e. something along the lines of (referencing my prior comment):

```ts
type Changes = {
  shard: Record<keyof Shard, ChangeField<Shard[keyof Shard]>>;
  feeds: Record<string, ChangeField<Feed>>;
  publishers: Record<string, ChangeField<Publisher>>;
}
```

In general, typecasting should be avoided, and if you have to use it for stuff like this, it generally indicates that you should rethink the types you're working with.